### ISSUE TITLE
Invalid iSCSI sessions handling when LUN is not available through all targets on portal

### ISSUE THREAD

**Issue by jkramarz** [xapi-project/sm#345]
Given iSCSI portal with multiple targets, where LUN is available only on one target, eg. targetd configured as follows:
```
{
  "fabric_modules": [],
  "storage_objects": [
    {
      "dev": "/var/lib/target/target1",
      "name": "var-lib-target-target1",
      "plugin": "fileio",
      "size": 10485760,
      "write_back": false,
      "wwn": "bf00cf22-a354-4a72-b51b-2d29637c26b8"
    }
  ],
  "targets": [
    {
      "fabric": "iscsi",
      "tpgs": [
        {
          "enable": true,
          "luns": [
            {
              "index": 0,
              "storage_object": "/backstores/fileio/var-lib-target-target1"
            }
          ],
          "node_acls": [
            {
              "mapped_luns": [
                {
                  "index": 0,
                  "tpg_lun": 0,
                  "write_protect": false
                }
              ],
              "node_wwn": "iqn.2012-06.com.example:initiator0"
            }
          ],
          "portals": [
            {
              "ip_address": "192.168.0.100",
              "iser": false,
              "port": 3260
            }
          ],
          "tag": 1
        }
      ],
      "wwn": "iqn.2012-06.com.example:target0"
    },
    {
      "fabric": "iscsi",
      "tpgs": [
          "luns": [],
          "node_acls": [
              "mapped_luns": [],
              "node_wwn": "iqn.2012-06.com.example:initiator0"
            }
          ],
          "portals": [
            {
              "ip_address": "192.168.0.100",
              "iser": false,
              "port": 3260
            }
          ],
          "tag": 1
        }
      ],
      "wwn": "iqn.2012-06.com.example:target2"
    }
  ]
}
```
[full configuration JSON](https://gist.github.com/jkramarz/0cccd6e9c71a41b38eb1536ab55e037f)

This configuration emulates behavior of [Dell Compellent storage](http://discussions.citrix.com/topic/384581-xenserver7-and-dell-compellent-issues-sc4020/).

When adding SR with wildcard Target IQN for multipathing support, operation fails with
`The SR is not available [opterr=Error reporting error, unknown key Device not appeared yet]` and `LUN offline or iscsi path down` due to incorrect iSCSI sessions handing.

Manually reconnecting (logging in) sessions results in operation success.
Following (completely incorrectly implemented) patch "fixes" this issue, but highly affects storage operations execution time:
```
diff --git a/drivers/LVHDoISCSISR.py b/drivers/LVHDoISCSISR.py
index 7562f36..ed4a439 100755
--- a/drivers/LVHDoISCSISR.py
+++ b/drivers/LVHDoISCSISR.py
@@ -145,6 +145,7 @@ class LVHDoISCSISR(LVHDSR.LVHDSR):
                     srcmd_copy.dconf['multiSession'] = IQNstring
                     util.SMlog("Setting targetlist: %s" % srcmd_copy.dconf['targetlist'])
                     self.iscsiSRs.append(BaseISCSI.BaseISCSISR(srcmd_copy, sr_uuid))
+                util.doexec(['iscsiadm', '-m', 'node', '--loginall=all'])
                 pbd = util.find_my_pbd(self.session, self.host_ref, self.sr_ref)
                 if pbd <> None and not self.dconf.has_key('multiSession'):
                     dconf = self.session.xenapi.PBD.get_device_config(pbd)
```

**Comment by germanop** [comment]
Hi @jkramarz ,
Thanks for reporting the issue.

We had in the past problems with dual controller arrays where a specific LUN was available on one controller but not on the other one (the other one was presented only for fail-over).

I am sure that some kind of problem could have arisen again (we did some refactoring and updated some userpsace tools) so, it would be interesting to repro your problem.
Can you propose a repro, unless the json you posted is sufficient?

We have some real hw for internal testing but if we could be exactly on the same page, it would be much better.

The proposed solution (as you stated) is not usable straight away and, anyway, we refcount the connections we bring up.
Nonetheless, we surely want to support as many iscsi configurations as possible and Compellent is already among them

**Comment by jkramarz** [comment]
Hi @germanop,

One clean XenServer 7.0.0 instance as an initiator and bare CentOS 7 with that configuration as iSCSI target will trigger the issue. I've added to [the gist with configuration](https://gist.github.com/jkramarz/0cccd6e9c71a41b38eb1536ab55e037f) full SMlog.

It seems that the exact conditions are:
- there're at last two targets returned by portals (one portal with two targets or two portals with one target on each),
- at last one target must have exactly 0 LUNs available
It simply fails, storage is declared as broken, sometimes may be "repaired", but for sure won't connect automatically after reboot.

Another interesting scenario would be replacing the second condition with "none of the LUNs are available on all targets", but I have not investigated it deeply.

For something like 2 weeks more I can provide you with any (even quite destructive) debugs from connecting real SCv2020 with latest firmware if you have only fully qualified ALUA storage in lab.

**Comment by germanop** [comment]
OK, thanks.

From what you say, it looks like a regression to me.
This kind of behaviour was fixed surely in 6.1.

I will check if what you reported is the same as something we are investigating internally
but any feedback or help is welcome.

**Comment by jkramarz** [comment]
I've tested it on sm-1.12.2-xs203 and sm-1.17.0 built from v1.17.0 tag with the same result.

**Comment by germanop** [comment]
On which XS release?

**Comment by jkramarz** [comment]
7.0.0-125380c with public updates up to XS70E027 applied.

**Comment by jkramarz** [comment]
I can't shake the feeling that fixing properly this problem also solves following issues:
- https://bugs.xenserver.org/browse/XSO-567
- https://bugs.xenserver.org/browse/XSO-678
- https://bugs.xenserver.org/browse/XSO-679
Would you mind sharing a bit more details on issue being investigated or progress? I'd rather prefer to not duplicate work you're doing internally, but willing to help (at last in finding the primary cause).

**Comment by germanop** [comment]
OK, I did not read all the posts but I saw a lot of similarities with what I saw.
Fixing the scsi_id command is one of them, even though does not explain why it does not happen
for the master (that's why I did not pushed the patch yet).
Master and slaves take different code paths for some reason.

I can send you an RPM for testing (or a patch, if you prefer) and you can try it.
It is just a few logs, a longer timeout and the scsi_id command fixed.

**Comment by jkramarz** [comment]
I'd be glad to try it out. Both - path or RPM - would do the job.
My email address is lenwe@lenwe.net .

**Comment by jkramarz** [comment]
I've applied it over sm-1.17.0.x1 and it works, thanks!

Not sure if I understand correctly how XAPI works, but assuming that when adding new SR, LUN discovery performed on pool master, I've noticed that after scanning for target LUNs there's residual, orphaned iSCSI session left to the target without LUNs. For me the following code may not be very careful about them, but maybe it only looks a bit hacky:
{code}
        # Apart from the upgrade case, user must specify a SCSIid
        if not self.dconf.has_key('SCSIid'):
            # Dual controller issue
            self.LUNs = {}  # Dict for LUNs from all the iscsi objects
            for ii in range(0, len(self.iscsiSRs)):
                self.iscsi = self.iscsiSRs[ii]
                self._LUNprint(sr_uuid)
                for key in self.iscsi.LUNs:
                    self.LUNs[key] = self.iscsi.LUNs[key]
            self.print_LUNs_XML()
            self.iscsi = self.iscsiSRs[0] # back to original value
            raise xs_errors.XenError('ConfigSCSIid')
{code}
Maybe it does make the difference?

**Comment by germanop** [comment]
This is for the probe phase but it is contained in the "create" call (don't ask me why :-) ).
Raising the exception is normal because this is how SM tells xapi that parameters are missing.
Whatever session is open during this phase is refcounted and, because of the exception, before
exiting, a cleanup is invoked and it is supposed to tear down all the open sessions.

I noticed, however, that this is not the case for this bug (but I am not worked on that, yet).

**Comment by germanop** [comment]
About the (rudimentary) patch, may I ask you to send me SMlog (both master and slave)
with the additional info about the time you attached the SR, so I can check the logs?

If you had only one host, would you have any problem, regardless of the patch?

**Comment by jkramarz** [comment]
I've just removed ("forgot") all remote storage on a pool of 3 servers (connected to Dell Compellent) and a single host (connected to a host configured as in issue description), rotated SMlog and re-attached one LUN. 

On one host I do had a problem, that got solved by this patch too.

**Comment by jkramarz** [comment]
No, it does not work without this line, but changes in `mpath_dmp.py` can be reverted.

**Comment by germanop** [comment]
OK, you are testing with multipath off, right?

**Comment by jkramarz** [comment]
It's inactive on single host, but pool has multipath enabled and 2 paths to the filer.

**Comment by germanop** [comment]
OK, so you tried on single host, with multipath disabled and this is where your logs are coming from, right?

**Comment by jkramarz** [comment]
The file "SMlog-xenserver1-without-LVHDoISCSISR-patch" (and all with xenserver in name) is from a host which is not in pool, has multipathing disabled, but with "*" as Target IQN and single iSCSI Target address.
In this setup I'm using configuration like in issue description, `iqn.2012-06.com.example:target0` returns 1 LUN and `iqn.2012-06.com.example:target2` returns 0 LUNs.

The files "SMlog-xen0{2,3,4}" are from a pool of 3 hosts with multipathing enabled and I'm providing two iSCSI Target addresses, resulting in 4 iSCSI sessions with 2 possible paths.

**Comment by germanop** [comment]
While I try to use the configuration you passed me (I need to check if things can be sped up a bit),
track the following that is what fixes things for you
https://github.com/xapi-project/sm/pull/348

**Comment by jkramarz** [comment]
Fixed in #348.
